doc/reference/kernel/smp/smp.rst

```diff
@@ -0,0 +1,292 @@
+.. _smp_arch:
+
+Symmetric Multiprocessing
+#########################
+
+On multiprocessor architectures, Zephyr supports the use of multiple
+physical CPUs running Zephyr application code.  This support is
+"symmetric" in the sense that no specific CPU is treated specially by
+default.  Any processor is capable of running any Zephyr thread, with
+access to all standard Zephyr APIs supported.
+
+No special application code needs to be written to take advantage of
+this feature.  If there are two Zephyr application threads runnable on
+a supported dual processor device, they will both run simultaneously.
+
+SMP configuration is controlled under the :option:`CONFIG_SMP` kconfig
+variable.  This must be set to "y" to enable SMP features, otherwise
+a uniprocessor kernel will be built.  In general the platform default
+will have enabled this anywhere it's supported. When enabled, the
+number of physical CPUs available is visible at build time as
+:option:`CONFIG_MP_NUM_CPUS`.  Likewise, the default for this will be the
+number of available CPUs on the platform and it is not expected that
+typical apps will change it.  But it is legal and supported to set
+this to a smaller (but obviously not larger) number for special
+purposes (e.g. for testing, or to reserve a physical CPU for running
+non-Zephyr code).
+
+Synchronization
+***************
+
+At the application level, core Zephyr IPC and synchronization
+primitives all behave identically under an SMP kernel.  For example
+semaphores used to implement blocking mutual exclusion continue to be
+a proper application choice.
+
+At the lowest level, however, Zephyr code has often used the
+``irq_lock()``/``irq_unlock()`` primitives to implement fine grained
+critical sections using interrupt masking.  These APIs continue to
+work via an emulation layer (see below), but the masking technique
+does not: the fact that your CPU will not be interrupted while you are
+in your critical section says nothing about whether a different CPU
+will be running simultaneously and be inspecting or modifying the same
+data!
+
+Spinlocks
+=========
+
+Instead, SMP systems provide a more constrained spinlock primitive:
```

Comment:
Starting this section with "Instead" doesn't real well.  Let's change this intro to:
```
SMP systems provide a more constrained ``k_spin_lock()``
primitive that not only masks interrupts locally, as done by
``irq_lock()``, but also atomically validates that a  ...
```